nomad/leader_test.go

```diff
@@ -629,6 +632,186 @@ func TestLeader_RestoreVaultAccessors(t *testing.T) {
 	}
 }
 
+func TestLeader_ClusterID(t *testing.T) {
+	t.Parallel()
+
+	s1 := TestServer(t, func(c *Config) {
+		c.NumSchedulers = 0
+		c.Build = minClusterIDVersion.String()
+	})
+	defer s1.Shutdown()
+	testutil.WaitForLeader(t, s1.RPC)
+
+	clusterID, err := s1.ClusterID()
+
+	require.NoError(t, err)
+	require.True(t, helper.IsUUID(clusterID))
+}
+
+func TestLeader_ClusterID_upgradePath(t *testing.T) {
+	t.Parallel()
+
+	before := version.Must(version.NewVersion("0.10.1")).String()
+	after := minClusterIDVersion.String()
+
+	outdated := func(bootstrap bool) *Server {
+		return TestServer(t, func(c *Config) {
+			c.NumSchedulers = 0
+			c.Build = before
+			c.DevDisableBootstrap = bootstrap
+			c.BootstrapExpect = 3
+			c.Logger.SetLevel(hclog.Trace)
+		})
+	}
+
+	upgraded := func(bootstrap bool) *Server {
+		return TestServer(t, func(c *Config) {
+			c.NumSchedulers = 0
+			c.Build = after
+			c.DevDisableBootstrap = bootstrap
+			c.BootstrapExpect = 3
+			c.Logger.SetLevel(hclog.Trace)
+		})
+	}
+
+	servers := []*Server{outdated(false), outdated(true), outdated(true)}
+	// fallback shutdown attempt in case testing fails
+	defer servers[0].Shutdown()
+	defer servers[1].Shutdown()
+	defer servers[2].Shutdown()
+
+	upgrade := func(i int) {
+		previous := servers[i]
+
+		servers[i] = upgraded(true)
+		TestJoin(t, servers[i], servers[(i+1)%3], servers[(i+2)%3])
+		testutil.WaitForLeader(t, servers[i].RPC)
+
+		require.NoError(t, previous.Leave())
+		require.NoError(t, previous.Shutdown())
+	}
+
+	// Join the servers before doing anything
+	TestJoin(t, servers[0], servers[1], servers[2])
+
+	// Wait for servers to settle
+	for i := 0; i < len(servers); i++ {
+		testutil.WaitForLeader(t, servers[i].RPC)
+	}
+
+	// A check that ClusterID is not available yet
+	noIDYet := func() {
+		for _, s := range servers {
+			retry.Run(t, func(r *retry.R) {
+				_, err := s.ClusterID()
+				require.Error(t, err)
```

Comment:
Shouldn't use `require` or `t` methods within `retry.Run`. I'd actually prefer not to use that library due to that gotcha, but it is convenient.

Reply:
eeesh yeah, good catch